doc/life-of-a-request.md

```diff
@@ -0,0 +1,164 @@
+
+### Protocol determination
+
+If protocol has not been pre-determined by policy, the proxy attempts to discover the
+protocol by reading data from the client. Note that this detection can break
+server-speaks-first protocols like MySQL and SMTP, necessitating the need to disable such
+detection by policy.
+
+In order to detect the connection's session protocol transparently (i.e. without policy),
+the proxy reads from the connection, buffering data until the proxy determines how to
+handle the connection.
+
+#### TLS
+
+If the connection begins with a TLS[[1]](#tls-q1) client handshake, data is read until
+until a full `ClientHello` message is buffered. The proxy determines whether it should
+terminate the TLS connection or treat the connection as an opaque TCP stream. The proxy
+may terminate TLS when a Conduit-specific extension[[2]](#tls-q2) exists in the client's
+hello, or it may terminate TLS if it is configured with a certificate for the
+client-requested name (i.e. for external/ingress).
+
+If the proxy elects to terminate TLS, the client's ALPN[[3]](#tls-q3) advertisement is
+used to determine which protocol should be used. If the ALPN advertisement does not
+include a proxy-supported protocol (i.e. `http` or `h2`), then the proxy handles the
+decrypted stream as opaque TCP[4](#tls-q4).
+
+##### Questions
+
+<a name="tls-q1"></a>
+1. What's the minimum version of the TLS protocol we support? I assume we don't handle SSL
+    3.0, etc. due to ClientHello requirement
+<a name="tls-q2"></a>
+2. How do TLS extensions play into this detection? are they conduit-specific or are we
+   reusing existing public extensions?
+<a name="tls-q3"></a>
+3. Can we reasonably expect clients to support ALPN? Do we have to support detecting
+   protocols for TLS requests that don't advertise a protocol?
+<a name="tls-q4"></a>
+4. It's not immediately clear to me how the proxy can "bridge" two TLS
+   connections--propagating ALPN information could be complex: we might want to issue a
+   `ClientHello` to the destination before issuing a ServerHello to the initiatior. This is
+   mostly germane for ingress.
+
+
+#### HTTP
+
+If the connection begins with an HTTP/1 request-line, the connection is handled as HTTP/1.
+
+If the connection begins with a HTTP/2 Prior-knowledge preface, the connection is handled
+as HTTP/2.
+
+
+## TCP Forwarding
+
+If, after protocol detection (and potentially TLS decryption), the connection's protocol
+is not supported by the proxy, the connection is simply forwarded to the original
+destination.
+
+Policy may require that the forwarded connections are TLS'd by the proxy (i.e. for mutual
+authentication).
+
+
+## HTTP Proxying
+
+### HTTP/1.0
+
+If the request is an HTTP/1.0 request that does not include a `Host` header, the request
+is routed to the connection's original destination address. The request is sent on a
+dedicated one-off connection (in keeping with the HTTP/1.0 specification).
```

Comment:
We need to handle absolute URLs in the request line too, not just the Host header.

Reply:
Yep, and in all HTTP/1.x cases, an absolute-form request target trumps any value in the `Host` header.

Reply:
Is there a term from the HTTP spec (preferably) or of our own that we can use to generalize "Request URI authority or Host header"? "Request authority"?

Reply:
RFC7230 tends to use the term [request-target](https://tools.ietf.org/html/rfc7230#section-5.3).